### PR TITLE
evdev-rs: add func to create Device using filename

### LINKED ISSUE #59: Open files with O_NONBLOCK
The documentation of libevdev states
```
* libevdev does not handle the file descriptors directly, it merely uses
* them. The caller is responsible for opening the file descriptors, setting
* them to `O_NONBLOCK` and handling permissions. A caller should drain any
* events pending on the file descriptor before passing it to libevdev.
 ```
(this is in `libevdev.h`)

Rust's `File::open` does not set `O_NONBLOCK` (it uses `O_RDONLY|O_CLOEXEC`). I'm not sure what problems this could cause.

~One way to fix this is to add a pub fn called `Device::open_from_path()`, which calls into libc directly.~ See comment below for code that uses only safe rust
<details>
<summary> unsafe example code (do not use) </summary>
 
```rust
pub fn open_from_path<P>(path: P) -> io::Result<Device>
where P: AsRef<OsStr>
{
    let path = CString::new(path.as_ref().as_bytes()).unwrap();
    let fd = match unsafe {
        libc::open(path.into_raw(), libc::O_RDONLY | libc::O_NONBLOCK | O_CLOEXEC)
    } {
        -1 => return Err(io::Error::last_os_error()),
        res => res
    };
    // maybe drain events here?
    Device::new_from_file(unsafe{File::from_raw_fd(fd)}))
}
```
</details>
(note: a way to improve this function is to use `fd`s directly to create a Device instead of rust `File`s)

Before we do this, I think we should investigate what if any bugs result from not opening with O_NONBLOCK

### PR DESCRIPTION
**Pull request by ndesh26**
Also documentation to highlight that the file for device
should always be opened with O_NONBLOCK flag set.

closes #59

Signed-off-by: Nayan Deshmukh <nayan26deshmukh@gmail.com>

### CODE REVIEW

File: examples/evtest.rs [hunk 1]
```diff
@@ -126,10 +129,20 @@ fn main() {
...
     let path = &args.nth(1).unwrap();
-    let f = File::open(path).unwrap();
+    let mut file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .custom_flags(libc::O_NONBLOCK)
+        .open(path)
+        .unwrap();
+    let mut buffer = Vec::new();
+    let result = file.read_to_end(&mut buffer);
```
Comment (maxbla) [inline examples/evtest.rs:139]: This potentially allocates (maybe even allocates a lot) for devices with many unread events. Consider using a fixed size buffer and a loop 
``` rust
let mut buf = [0u8; std::mem::size_of::<input_event>()*20];
while file.read(&mut buf).is_ok() {}
```
or something with file.bytes()

**Comment by maxbla** [comment]
The examples (evtest, vmouse) were updated to use `O_NONBLOCK`, but they would be shorter and have equivalent code if they instead used `new_from_path`. I think the tests should be changed too.

File: examples/evtest.rs (hunk 1 above)
Comment (ndesh26) [reply examples/evtest.rs:139]: Reading in fixed-size sounds better. I'll update the PR.

**Comment by ndesh26** [comment]
> The examples (evtest, vmouse) were updated to use `O_NONBLOCK`, but they would be shorter and have equivalent code if they instead used `new_from_path`. I think the tests should be changed too.

Updating tests sounds fine. But I wanted to leave at least one use of new_from_file which is properly setting the flag and draining the events for a user to refer to. I feel that it might go unnoticed in the docs.

File: src/device.rs [hunk 2]
```diff
@@ -656,9 +656,16 @@ impl Device {
...
             .custom_flags(libc::O_NONBLOCK)
             .open(path)?;
-        let mut buffer = Vec::new();
+        let mut buffer = [0u8; 20 * std::mem::size_of::<raw::input_event>()];
+        let mut result;
+        loop {
+            result = file.read(&mut buffer);
+            if !result.is_ok() {
+                break;
+            };
```
Comment (maxbla) [inline src/device.rs:665]: nit: prefer fewer mutable variables
``` rust
let first_error = loop {
    let result = file.read(&mut buffer);
    if !result.is_ok() {
        break result;
    }
};

match first_error {
//...
```

File: src/device.rs [hunk 3]
```diff
@@ -617,6 +621,10 @@ impl Device {
     /// # let file = File::open("/dev/input/event0").unwrap();
     /// let device = uninit_device.set_file(file);
     /// ```
+    /// The caller is responsible for opening the file and setting
+    /// the `O_NONBLOCK` flag and handling permissions.
+    /// If the file is opened without O_NONBLOCK flag then next_event
+    /// should be called with ReadFlag::BLOCKING.
```
Comment (maxbla) [inline src/device.rs:627]: I would add something like:

> If the file is not opened with `O_NONBLOCK` and `next_event()` is not called with `ReadFlag::BLOCKING`, then a call to `next_event()` may block trying to buffer events that have not yet happened.

so it is clear why those flags should be used

File: src/device.rs [hunk 4]
```diff
@@ -641,6 +649,31 @@ impl Device {
         Self::new_from_file(file)
     }
 
+    /// Opens a device with the given path as the location of devnode
```
Comment (maxbla) [inline src/device.rs:652]: document that this method also reads off all pending events

File: src/device.rs (hunk 3 above)
Comment (ndesh26) [reply src/device.rs:627]: Yes it would be good to capture this information

**Pull request merged**